project/profiles/capacity4more/modules/c4m/og/c4m_og_invite_visitors/c4m_og_invite_visitors.module

```diff
@@ -0,0 +1,671 @@
+  if (!password_verify($token, $invitation->inv_token)) {
+    drupal_goto($group->purl);
+  }
+
+  // An invitation that had been processed before.
+  if ($invitation->inv_expired) {
+    drupal_set_message(t('This invitation link has expired.'), 'warning');
+    drupal_goto($group->purl);
+  }
+
+  // A user is already logged in.
+  if ($user->uid) {
+    // If the logged in email is the same as the invitation email, add to group.
+    // Otherwise, display an error.
+    if ($user->mail == $email) {
+      // Check if the user is a member already. Add if not.
+      if (og_is_member('node', $group->nid, 'user', $user)) {
+        drupal_goto($group->purl);
+      }
+      else {
+        // Add the user to the group.
+        $values = array(
+          'entity_type' => 'user',
+          'entity' => $user,
+          'state' => OG_STATE_ACTIVE,
+        );
+        og_group('node', $group, $values);
+      }
+    }
+    else {
+      drupal_set_message('Another user is logged in. Please log out and try again.');
+      drupal_goto($group->purl);
+    }
+  }
+
+  // Everything is Kosher, proceed to the registration form.
+  // Set group to join in session variable and redirect to /user/register.
+  $_SESSION['c4m_og_invite_visitors_gid_join'] = $group->nid;
+  drupal_goto('/user/register');
+}
+
+/**
+ * Implements hook_og_ui_get_group_admin().
+ *
+ * Adds a menu item (Invite visitors) to the group administration page. Links to
+ * the invite visitors form.
+ */
+function c4m_og_invite_visitors_og_ui_get_group_admin($group_type, $gid) {
+  $items = array();
+  if (og_user_access($group_type, $gid, 'invite visitors')) {
+    $items['c4m_og_invite_visitors'] = array(
+      'title' => t('Invite visitors'),
+      'description' => t('Create new users and add them as group members.'),
+      'href' => 'admin/people/invite-visitors',
+    );
+  }
+
+  return $items;
+}
+
+/**
+ * Invite access. Checks if the user is allowed to invite visitors to the group.
+ *
+ * Checks by user role and og_role against group visibility (Public/Private) and
+ * Membership requests setting (Open/Moderated).
+ *
+ * @param int $gid
+ *   The group #ID from the URL.
+ *
+ * @return bool
+ *   Grand/deny access for inviting visitors.
+ */
+function c4m_og_invite_visitors_user_access($gid) {
+  global $user;
+
+  $group = node_load($gid);
+  if (!$group || !og_is_group('node', $group)) {
+    return FALSE;
+  }
+
+  // OG permission.
+  if (!og_user_access('node', $gid, 'invite visitors')) {
+    return FALSE;
+  }
+
+  // SAs, GOs and GAs can invite new members.
+  if (_c4m_features_og_members_is_power_user($group)) {
+    return TRUE;
+  }
+
+  // Get request type (open/moderated).
+  $membership_request_value = c4m_og_get_group_membership_request($group);
+  // Get access type (public/private).
+  $access_type = c4m_og_get_access_type($group);
+
+  // For private or moderated groups, return false. Else check for member.
```

Comment:
=> `// Non power users are not allowed to invite visitors to a private / moderated group.`